addons/web/static/src/core/domain.js

```diff
@@ -230,9 +230,15 @@ function normalizeDomainAST(domain, op = "&") {
     }
     let expected = 1;
     for (const child of domain.value) {
-        if (child.type === 1 /* String */ && (child.value === "&" || child.value === "|")) {
-            expected++;
-        } else if (child.type !== 1 /* String */ || child.value !== "!") {
+        if (child.type === 1 /* String */) {
+            if (child.value === "&" || child.value === "|") {
+                expected++;
+            } else if (child.value !== "!") {
+                throw new InvalidDomainError("Invalid domain AST");
+            }
+        } else if (![4, 10].includes(child.type) || child.value.length !== 3) {
```

Comment:
i think it would be clearer to use a switch statement, and also to indicate what the magic numbers are. Something like this: 

```js
for (const child of domain.value) {
  switch (child.type) {
    case 1: /* String */ 
            if (child.value === "&" || child.value === "|") {
                expected++;
            } else if (child.value !== "!") {
                throw new InvalidDomainError("Invalid domain AST");
            }
    case 4:  /* list */
    case 10: /* tuple */
        if (child.value.length === 3) {
           expected--;
           break;
       }
    default:
            throw new InvalidDomainError("Invalid domain AST");


  }
}
```

Reply:
it would be nicer indeed. Updated